### ISSUE TITLE
Anchor Optimization

### ISSUE THREAD

**Issue by lleye** [ultralytics/yolov5#3514]
## ❔Question
how do i turn on anchor optimization. I know it is enabled by default but it seems it only works when i set the 
anchors: # to a value in the hyp file, if i omit it then or comment it out, autoanchor is not run. is this by design?

@glenn-jocher

**Comment by glenn-jocher** [comment]
👋 Hello! Thanks for asking about model anchors. YOLOv5 🚀 uses a new [Ultralytics](https://ultralytics.com/) algorithm called **autoanchor** for anchor verification and generation before training starts.

Autoanchor will analyse your anchors against your dataset and training settings (like `--img-size`), and will adjust your anchors as necessary if it determines the original anchors are a poor fit, or if an **anchor count** was specified in your model.yaml rather than **anchor values**, i.e. 
```yaml
# Specify anchor count (per layer)
anchors: 3

# --OR-- Specify anchor values manually
anchors:
  - [10,13, 16,30, 33,23]  # P3/8
  - [30,61, 62,45, 59,119]  # P4/16
  - [116,90, 156,198, 373,326]  # P5/32
```

When generating new anchors, autoanchor first applies a kmeans function against your dataset labels (scaled to your training `--img-size`), and then uses these kmeans centroids as initial conditions for a Genetic Evolution (GE) algorithm. The GE algorithm will evolve all anchors for 1000 generations under default settings, using the CIoU loss (same one used during training) combined with Best Possible Recall (BPR) as its fitness function.

No action is required on your part to use autoanchor. If you would like to force manual anchors for any reason, you can skip autoanchor with the `--noautoanchor` flag:
```shell
python train.py --noautoanchor
```

**Comment by wmcnally** [comment]
@glenn-jocher have you experimented with using more anchor boxes per output layer (e.g., 4 or 5)? Why is 3 the default?

**Comment by glenn-jocher** [comment]
@wmcnally all hyperparameters are determined via evolution.

**Comment by Minh-Tu-Cao** [comment]
> @wmcnally all hyperparameters are determined via evolution.

Can you tell me the coding section where GA is used for optimize anchor boxes? Thank you

**Comment by wmcnally** [comment]
@Minh-Tu-Cao https://github.com/ultralytics/yolov5/blob/master/train.py#L235

**Comment by glenn-jocher** [comment]
@Minh-Tu-Cao autoanchor.py contains the anchor evolution code:
https://github.com/ultralytics/yolov5/blob/fc3606420d86fac54b6fe5e67fc4be64325df5dc/utils/autoanchor.py#L146-L162

**Comment by Minh-Tu-Cao** [comment]
Got it! Thank you @glenn-jocher

**Comment by Minh-Tu-Cao** [comment]
> @Minh-Tu-Cao https://github.com/ultralytics/yolov5/blob/master/train.py#L235

Thank you @wmcnally
